### PR TITLE
Tibbles

### PR DESCRIPTION
**Pull request by maelle**
cf #43 

* the code is very ugly

* it does not work for latest yet

Is this the kind of tibbles you had in mind @gaborcsardi?

``` r
library("pkgsearch")
cran_packages(format = "short")
#> # A tibble: 10 x 2
#>    version title                                                           
#>  * <chr>   <chr>                                                           
#>  1 1.0.0   "Accurate, Adaptable, and Accessible Error Metrics for Predicti…
#>  2 0.5.4   Access to Abbyy Optical Character Recognition (OCR) API         
#>  3 2.1     Tools for Approximate Bayesian Computation (ABC)                
#>  4 1.0     Data Only: Tools for Approximate Bayesian Computation (ABC)     
#>  5 0.9.0   Array Based CpG Region Analysis Pipeline                        
#>  6 1.2.1   Computed ABC Analysis                                           
#>  7 0.4     ABCDE_FBA: A-Biologist-Can-Do-Everything of Flux Balance<U+000a…
#>  8 0.15.0  Implementation of Artificial Bee Colony (ABC) Optimization      
#>  9 1.2     Approximate Bayesian Computational Model for Estimating P2      
#> 10 1.7.1   Approximate Bayesian Computation via Random Forests
cran_packages(format = "full")
#> # A tibble: 54 x 7
#>    name   versions  timeline latest title                  archived revdeps
#>  * <chr>  <list>    <list>   <chr>  <chr>                  <lgl>      <dbl>
#>  1 A3     <list [1… <chr [1… 1.0.0  "Accurate, Adaptable,… FALSE          0
#>  2 A3     <list [1… <chr [1… 1.0.0  "Accurate, Adaptable,… FALSE          0
#>  3 A3     <list [1… <chr [1… 1.0.0  "Accurate, Adaptable,… FALSE          0
#>  4 abbyyR <list [2… <chr [1… 0.5.4  Access to Abbyy Optic… FALSE          0
#>  5 abbyyR <list [2… <chr [1… 0.5.4  Access to Abbyy Optic… FALSE          0
#>  6 abbyyR <list [2… <chr [1… 0.5.4  Access to Abbyy Optic… FALSE          0
#>  7 abbyyR <list [2… <chr [1… 0.5.4  Access to Abbyy Optic… FALSE          0
#>  8 abbyyR <list [2… <chr [1… 0.5.4  Access to Abbyy Optic… FALSE          0
#>  9 abbyyR <list [2… <chr [1… 0.5.4  Access to Abbyy Optic… FALSE          0
#> 10 abbyyR <list [2… <chr [1… 0.5.4  Access to Abbyy Optic… FALSE          0
#> # … with 44 more rows
cran_packages(format = "latest")
#> Error in `[<-.data.frame`(`*tmp*`, col, value = c("2.15.2", "2.15.3", : replacement has 8 rows, data has 1
```

<sup>Created on 2019-05-16 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

### CODE REVIEW

**Comment by maelle** [comment]
I guess the tibble corresponding to the full format should be unnested a bit more (one line per version of a package, many more columns).

To make the code prettier maybe I should have one unnesting/wrangling function by format?

**Comment by gaborcsardi** [comment]
Yes, I think this is getting good. I am thinking that having a function that potentially returns different _kind of_ things is not such a good practice, and maybe we want three functions, actually. Or get rid of the short format, because that's a subset, anyway, and have two functions?

**Comment by maelle** [comment]
That makes sense. Now I wonder how to call them :thinking: 

`cran_packages()` for latest vs `cran_all_packages()` for full? It's not clear.

**Comment by gaborcsardi** [comment]
short: cran_package_list()
latest: cran_active_packages()
full: cran_package_histories()

They can return essentially the same format, but 
- 'short' will have only a subset of the columns,
- 'latest' will have all columns, and one row per package
- 'full' will have all columns and multiple rows per package

Maybe?

**Comment by maelle** [comment]
Yes makes sense, i like the use of the word history

**Comment by maelle** [comment]
@gaborcsardi here is the current output of all three functions. Ok? What if ok, you can review the code itself. You'll see I unstuck myself (from what I experienced as unnesting hell) by copy-pasting code from `desc` (MIT licence too :tada:).

Once you're happy with the output and the code, I'll make the tests more thorough. :angel: 

I do not understand the revdeps field. Should it just be removed? Related to #37 I assume.

``` r
library("pkgsearch")

cran_package_list()
#> # A tibble: 10 x 2
#>    version title                                                           
#>  * <chr>   <chr>                                                           
#>  1 1.0.0   "Accurate, Adaptable, and Accessible Error Metrics for Predicti…
#>  2 0.5.4   Access to Abbyy Optical Character Recognition (OCR) API         
#>  3 2.1     Tools for Approximate Bayesian Computation (ABC)                
#>  4 1.0     Data Only: Tools for Approximate Bayesian Computation (ABC)     
#>  5 0.9.0   Array Based CpG Region Analysis Pipeline                        
#>  6 1.2.1   Computed ABC Analysis                                           
#>  7 0.4     ABCDE_FBA: A-Biologist-Can-Do-Everything of Flux Balance<U+000a…
#>  8 0.15.0  Implementation of Artificial Bee Colony (ABC) Optimization      
#>  9 1.2     Approximate Bayesian Computational Model for Estimating P2      
#> 10 1.7.1   Approximate Bayesian Computation via Random Forests

cran_active_packages()
#> # A tibble: 10 x 31
#>    Package Type  Title Version Date  Author Maintainer Description License
#>  * <chr>   <chr> <chr> <chr>   <chr> <chr>  <chr>      <chr>       <chr>  
#>  1 A3      Pack… "Acc… 1.0.0   2015… Scott… Scott For… Supplies t… GPL (>…
#>  2 abbyyR  <NA>  Acce… 0.5.4   <NA>  Gaura… Gaurav So… "Get text … MIT + …
#>  3 abc     Pack… Tool… 2.1     2015… "Csil… Blum Mich… "Implement… GPL (>…
#>  4 abc.da… Pack… Data… 1.0     2015… "Csil… Blum Mich… Contains d… GPL (>…
#>  5 ABC.RAP <NA>  Arra… 0.9.0   <NA>  Abdul… Abdulmone… "It aims t… GPL-3  
#>  6 ABCana… Pack… Comp… 1.2.1   2017… Micha… Florian L… "For a giv… GPL-3  
#>  7 abcdeF… Pack… ABCD… 0.4     2012… Abhil… Abhilash … Functions … GPL-2  
#>  8 ABCopt… Pack… Impl… 0.15.0  2017… "Geor… George Ve… "An implem… MIT + …
#>  9 ABCp2   Pack… Appr… 1.2     2016… M. Ca… M. Cather… Tests the … GPL-2  
#> 10 abcrf   Pack… Appr… 1.7.1   2018… "Jean… Jean-Mich… "Performs … GPL (>…
#> # … with 22 more variables: NeedsCompilation <chr>, Packaged <chr>,
#> #   Repository <chr>, `Date/Publication` <chr>, crandb_file_date <chr>,
#> #   date <chr>, dependencies <list>, `Authors@R` <chr>, URL <chr>,
#> #   BugReports <chr>, LazyData <chr>, VignetteBuilder <chr>,
#> #   RoxygenNote <chr>, MD5sum <chr>, Contact <chr>, Encoding <chr>,
#> #   SystemRequirements <chr>, LazyLoad <chr>, Lazyload <chr>,
#> #   releases <list>, `Classification/ACM` <chr>,
#> #   `Classification/JEL` <chr>

cran_package_histories()
#> # A tibble: 54 x 33
#>    Package Type  Title Version Date  Author Maintainer Description License
#>  * <chr>   <chr> <chr> <chr>   <chr> <chr>  <chr>      <chr>       <chr>  
#>  1 A3      Pack… A3: … 0.9.1   2013… Scott… Scott For… This packa… GPL (>…
#>  2 A3      Pack… A3: … 0.9.2   2013… Scott… Scott For… This packa… GPL (>…
#>  3 A3      Pack… "Acc… 1.0.0   2015… Scott… Scott For… Supplies t… GPL (>…
#>  4 abbyyR  <NA>  Acce… 0.1     <NA>  Gaura… Gaurav So… Get text f… GPL (>…
#>  5 abbyyR  <NA>  Acce… 0.2     <NA>  Gaura… Gaurav So… Get text f… GPL (>…
#>  6 abbyyR  <NA>  Acce… 0.2.1   <NA>  Gaura… Gaurav So… Get text f… MIT + …
#>  7 abbyyR  <NA>  Acce… 0.2.2   <NA>  Gaura… Gaurav So… Get text f… MIT + …
#>  8 abbyyR  <NA>  Acce… 0.2.3   <NA>  Gaura… Gaurav So… Get text f… MIT + …
#>  9 abbyyR  <NA>  Acce… 0.3     <NA>  Gaura… Gaurav So… Get text f… MIT + …
#> 10 abbyyR  <NA>  Acce… 0.4.0   <NA>  Gaura… Gaurav So… Get text f… MIT + …
#> # … with 44 more rows, and 24 more variables: Packaged <chr>,
#> #   Repository <chr>, `Date/Publication` <chr>, crandb_file_date <chr>,
#> #   date <chr>, releases <list>, dependencies <list>,
#> #   NeedsCompilation <chr>, archived <lgl>, revdeps <dbl>,
#> #   `Authors@R` <chr>, URL <chr>, BugReports <chr>, LazyData <chr>,
#> #   VignetteBuilder <chr>, RoxygenNote <chr>, MD5sum <chr>, Contact <chr>,
#> #   Encoding <chr>, SystemRequirements <chr>, LazyLoad <chr>,
#> #   Lazyload <chr>, `Classification/ACM` <chr>, `Classification/JEL` <chr>
cran_package_histories(from = "ggplot2")$revdeps
#>  [1] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
#> [36] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
#> [71] 0 0 0
```

<sup>Created on 2019-05-23 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

File: DESCRIPTION [hunk 1]
```diff
@@ -24,7 +24,7 @@ Imports:
     dotenv
 Suggests:
     covr,
-    testthat,
+    testthat (>= 2.1.0),
```
Comment (maelle) [inline DESCRIPTION]: this got added by `usethis::use_test()`

**Comment by gaborcsardi** [comment]
Thanks for working on this, and removed some entry points that are too slow. I refactored a bunch of things, and will merge this now. We can still polish it later.

**Pull request merged**